piker/brokers/binance.py

```diff
@@ -0,0 +1,496 @@
+        sym: Optional[str] = None
+
+    ) -> dict:
+
+        resp = await self._api('exchangeInfo', {})
+        if sym is not None:
+            for sym_info in resp['symbols']:
+                if sym_info['symbol'] == sym:
+                    return sym_info
+            else:
+                raise SymbolNotFound(f'{sym} not found')
+        else:
+            return resp['symbols']
+
+    async def bars(
+        self,
+        symbol: str = 'BTCUSDT',
+        start_time: int = None,
+        end_time: int = None,
+        limit: int = 1000,  # <- max allowed per query
+        as_np: bool = True,
+
+    ) -> dict:
+
+        if start_time is None:
+            start_time = binance_timestamp(
+                arrow.utcnow().floor('minute').shift(minutes=-limit)
+            )
+
+        if end_time is None:
+            end_time = binance_timestamp(arrow.utcnow())
+
+        # https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-data
+        bars = await self._api(
+            'klines',
+            {
+                'symbol': symbol,
+                'interval': '1m',
+                'startTime': start_time,
+                'endTime': end_time,
+                'limit': limit
+            }
+        )
+
+        # TODO: pack this bars scheme into a ``pydantic`` validator type:
+        # https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-data
+
+        # TODO: we should port this to ``pydantic`` to avoid doing
+        # manual validation ourselves..
+        new_bars = []
+        for i, bar in enumerate(bars):
+
+            bar = OHLC(*bar)
+
+            row = []
+            for j, (name, ftype) in enumerate(_ohlc_dtype[1:]):
+
+                # TODO: maybe we should go nanoseconds on all
+                # history time stamps?
+                if name == 'time':
+                    # convert to epoch seconds: float
+                    row.append(bar.time / 1000.0)
+
+                else:
+                    row.append(getattr(bar, name))
+
+            new_bars.append((i,) + tuple(row))
+
+        array = np.array(new_bars, dtype=_ohlc_dtype) if as_np else bars
+        return array
+
+
+@asynccontextmanager
+async def get_client() -> Client:
+    yield Client()
+
+
+# validation type
+class AggTrade(BaseModel):
+    e: str  # Event type
+    E: int  # Event time
+    s: str  # Symbol
+    a: int  # Aggregate trade ID
+    p: float  # Price
+    q: float  # Quantity
+    f: int  # First trade ID
+    l: int  # Last trade ID
+    T: int  # Trade time
+    m: bool  # Is the buyer the market maker?
+    M: bool  # Ignore
+
+
+async def stream_messages(ws):
+    while True:
+
+        with trio.move_on_after(5):
+            msg = await ws.recv_msg()
+
+        # for l1 streams binance doesn't add an event type field so
+        # identify those messages by matching keys
+        # https://binance-docs.github.io/apidocs/spot/en/#individual-symbol-book-ticker-streams
+
+        if msg.get('u'):
```

Comment:
Oh that was my other question: can we do similar `BaseModel` unpacking/validation here for `l1` messages?